pkg/file/redundancy/redundancy.go

```diff
@@ -0,0 +1,204 @@
+	maxShards  int        // number of chunks after which the parity encode function should be called
+	maxParity  int        // number of parity chunks if maxShards has been reached for erasure coding
+	encryption bool
+}
+
+func New(level Level, encryption bool, pipeLine pipeline.PipelineFunc) *Params {
+	maxShards := 0
+	maxParity := 0
+	if encryption {
+		maxShards = level.GetMaxEncShards()
+		maxParity = level.GetParities(swarm.EncryptedBranches)
+	} else {
+		maxShards = level.GetMaxShards()
+		maxParity = level.GetParities(swarm.BmtBranches)
+	}
+	// init dataBuffer for erasure coding
+	rsChunkLevels := 0
+	if level != NONE {
+		rsChunkLevels = maxLevel
+	}
+	Buffer := make([][][]byte, rsChunkLevels)
+	for i := 0; i < rsChunkLevels; i++ {
+		Buffer[i] = make([][]byte, swarm.BmtBranches) // 128 long always because buffer varies at encrypted chunks
+	}
+
+	return &Params{
+		level:      level,
+		pipeLine:   pipeLine,
+		buffer:     Buffer,
+		cursor:     make([]int, 9),
+		maxShards:  maxShards,
+		maxParity:  maxParity,
+		encryption: encryption,
+	}
+}
+
+// ACCESSORS
+
+func (p *Params) MaxShards() int {
+	return p.maxShards
+}
+
+func (p *Params) Level() Level {
+	return p.level
+}
+
+// METHODS
+
+func (p *Params) Parities(shards int) int {
+	if p.encryption {
+		return p.level.GetEncParities(shards)
+	}
+	return p.level.GetParities(shards)
+}
+
+// ChunkWrite caches the chunk data on the given chunk level and if it is full then it calls Encode
+func (p *Params) ChunkWrite(chunkLevel int, data []byte, callback ParityChunkCallback) error {
+	if p.level == NONE {
+		return nil
+	}
+	if len(data) != swarm.ChunkWithSpanSize {
+		zeros := make([]byte, swarm.ChunkWithSpanSize-len(data))
+		data = append(data, zeros...)
+	}
+
+	return p.chunkWrite(chunkLevel, data, callback)
+}
+
+// ChunkWrite caches the chunk data on the given chunk level and if it is full then it calls Encode
+func (p *Params) chunkWrite(chunkLevel int, data []byte, callback ParityChunkCallback) error {
+	// append chunk to the buffer
+	p.buffer[chunkLevel][p.cursor[chunkLevel]] = data
+	p.cursor[chunkLevel]++
+
+	// add parity chunk if it is necessary
+	if p.cursor[chunkLevel] == p.maxShards {
+		// append erasure coded data
+		return p.encode(chunkLevel, callback)
+	}
+	return nil
+}
+
+// Encode produces and stores parity chunks that will be also passed back to the caller
+func (p *Params) Encode(chunkLevel int, callback ParityChunkCallback) error {
+	if p.level == NONE || p.cursor[chunkLevel] == 0 {
+		return nil
+	}
+
+	return p.encode(chunkLevel, callback)
+}
+
+func (p *Params) encode(chunkLevel int, callback ParityChunkCallback) error {
+	shards := p.cursor[chunkLevel]
+	parities := p.Parities(shards)
+
+	n := shards + parities
+	// realloc for parity chunks if it does not override the prev one
+	// caculate parity chunks
+	enc, err := erasureEncoderFunc(shards, parities)
+	if err != nil {
+		return err
+	}
+	// make parity data
```

Comment:
This comment doesn't add much value. The same for most of the comments below. Comment the why, not the what, the idea being that the code should be self-documenting and the comments should only be a last resort for explaining stuff. Comments can get out of sync, but the code is always the code, so there's no need for unnecessary redundancy.